modules/gapi/src/streaming/onevpl/accelerators/surface/dx11_frame_adapter.cpp

```diff
@@ -0,0 +1,232 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2021 Intel Corporation
+
+#include "streaming/onevpl/accelerators/surface/dx11_frame_adapter.hpp"
+#include "streaming/onevpl/accelerators/dx11_alloc_resource.hpp"
+#include "streaming/onevpl/accelerators/surface/surface.hpp"
+#include "logger.hpp"
+
+#ifdef HAVE_ONEVPL
+#include "streaming/onevpl/onevpl_export.hpp"
+
+#ifdef HAVE_INF_ENGINE
+// For IE classes (ParamMap, etc)
+#include <inference_engine.hpp>
+#endif // HAVE_INF_ENGINE
+
+namespace cv {
+namespace gapi {
+namespace wip {
+namespace onevpl {
+
+void lock_mid(mfxMemId mid, mfxFrameData &data, MediaFrame::Access mode) {
+    LockAdapter* alloc_data = reinterpret_cast<LockAdapter *>(mid);
+    if (mode == MediaFrame::Access::R) {
+        alloc_data->read_lock(mid, data);
+    } else {
+        alloc_data->write_lock(mid, data);
+    }
+}
+
+void unlock_mid(mfxMemId mid, mfxFrameData &data, MediaFrame::Access mode) {
+    LockAdapter* alloc_data = reinterpret_cast<LockAdapter*>(data.MemId);
+    if (mode == MediaFrame::Access::R) {
+        alloc_data->unlock_read(mid, data);
+    } else {
+        alloc_data->unlock_write(mid, data);
```

Comment:
so can this be a plain abstract `rwlock` (e.g. like one from boost or pthreads)? Or something differs in the access at the VPL level?

Reply:
adapter is quite different, but shared_mutex inside it should be reused from official sources further
P.S. `is_own_locked`  can be put on upper level to do not customize  official shared_mutex
P.P.S. official c++17 shared_mutex is mutex-based with wait-lock, but tbb has imple,entation with rw_**spinlock** to make busy wait